alchemiscale_fah/compute/service.py

```diff
@@ -4,6 +4,621 @@
+        )
+        formatter.converter = time.gmtime  # use utc time for logging timestamps
+
+        sh = logging.StreamHandler()
+        sh.setFormatter(formatter)
+        logger.addHandler(sh)
+
+        if self.settings.logfile is not None:
+            fh = logging.FileHandler(self.settings.logfile)
+            fh.setFormatter(formatter)
+            logger.addHandler(fh)
+
+        self.logger = logging.LoggerAdapter(logger, extra)
+
+        self.heartbeat_thread = None
+        self.fah_cert_update_thread = None
+
+    def update_fah_cert(self):
+        """Start up the FAH cert update, sleeping for `self.fah_cert_update_interval`"""
+        while True:
+            if self._stop:
+                break
+            self.fah_client.as_update_certificate()
+            time.sleep(self.fah_cert_update_interval)
+
+    def _refresh_heartbeat_thread(self):
+        if self.heartbeat_thread is None:
+            self.heartbeat_thread = threading.Thread(target=self.heartbeat, daemon=True)
+            self.heartbeat_thread.start()
+
+        # check that heartbeat is still alive; if not, resurrect it
+        elif not self.heartbeat_thread.is_alive():
+            self.heartbeat_thread = threading.Thread(target=self.heartbeat, daemon=True)
+            self.heartbeat_thread.start()
+
+    def _refresh_cert_update_thread(self):
+        if self.fah_cert_update_thread is None:
+            self.fah_cert_update_thread = threading.Thread(
+                target=self.update_fah_cert, daemon=True
+            )
+            self.fah_cert_update_thread.start()
+
+        # check that heartbeat is still alive; if not, resurrect it
+        elif not self.fah_cert_update_thread.is_alive():
+            self.fah_cert_update_thread = threading.Thread(
+                target=self.update_fah_cert, daemon=True
+            )
+            self.fah_cert_update_thread.start()
+
+    async def async_execute(self, task: ScopedKey) -> ScopedKey:
+        """Executes given Task.
+
+        Returns ScopedKey of ProtocolDAGResultRef following push to database.
+
+        """
+        # check if Task seen before, serialized ProtocolDAG present
+        # use that ProtocolDAG instead and feed to `execute_DAG`
+        project_id, run_id, clone_id = self.index.get_task(task)
+
+        if project_id is not None:
+            protocoldag = self.index.get_task_protocoldag(task)
+        else:
+            # check if we have seen this Transformation before
+            # get PROJECT, RUN if so
+            tf_sk = self.client.get_task_transformation(task)
+            project_id, run_id = self.index.get_transformation(tf_sk.gufe_key)
+            clone_id = None
+
+            # obtain a ProtocolDAG from the task
+            self.logger.info("Creating ProtocolDAG from '%s'...", task)
+            protocoldag, transformation, extends = self.task_to_protocoldag(task)
+            self.logger.info(
+                "Created '%s' from '%s' performing '%s'",
```

Comment:
use fstring

Reply:
For logging, it's generally recommended to use the `%s` syntax over f-strings, since the latter are always evaluated but `%s` usage only evaluated by logger if the given loglevel applies.

Reply:
https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs